### PR TITLE
Flat earth

### PR DESCRIPTION
**Pull request by Jeremy-Gaillard**
This pull request offers a flat representation of the world, especially useful for working on small-scale scenes (such as a city).
It also features a WMS provider and the possibility to add a texture to a BasicMaterial. (Unfortunately, it seems another WMS Provider was pushed to the master a few hours ago).

This is still work in progress, but I hope to bring some attention to this branch. It is close to working properly, but I have some visibility issues when I dezoom or tilt the view. Changing the far plane doesn't solve the issue.

There is also a need to add a proper srs to each object (Node ? Tile ?), to move some code "higher" in the call stack (e.g. the configuration of the WMS provider shouldn't be in FlatTileProvider but in Scene) and to clean up some (dead) code. These tasks are however not specific to this pull request but to all the current code.

### CODE REVIEW

File: src/Scene/NodeProcess.js [hunk 1]
```diff
@@ -88,7 +88,7 @@ define('Scene/NodeProcess',['Scene/BoundingBox','Renderer/Camera','Core/Math/Mat
 
         var obb     = node.OBB();               
         
-        var l       = node.absoluteCenter.length();
+        var l       = 0 * node.absoluteCenter.length(); // TODO : is this really necessary? (note : bugs on plane mode if not 0)
```
Comment (Jeremy-Gaillard) [inline src/Scene/NodeProcess.js:84]: I didn't understand this part of the code. If I didn't put l = 0, my geometries wouldn't appear.

**Comment by gchoqueux** [comment]
after reflection, I think it is possible to have a flat globe in very little change

**Comment by nosy-b** [comment]
Ok super we must check the flat earth prototype you did and try the prototype with few modification (high level @gchoqueux  too).
I modified the existing WMS provider to be compatible with your version.

**Comment by Jeremy-Gaillard** [comment]
Any insight on the visibility problems I still have @gchoqueux ?

Edit: seems to be the far plane after all. Is it possible to work on a merge ?

**Comment by Jeremy-Gaillard** [comment]
I'm encountering problems with the depth buffer.
You can reproduce the bug in the second image by dezooming and tilting the camera a bit.
Do you have any idea of what can be causing this problem ? @gchoqueux 
depth1
depth2

**Comment by nosy-b** [comment]
Are you using the last master? The new elevation process is maybe in cause... We check

**Comment by Jeremy-Gaillard** [comment]
No I am not. There isn't even an elevation, it's just textured flat tiles with a skirt.

**Comment by Jeremy-Gaillard** [comment]
Ok, we solved the problem with @qdnguyen : the infinite skirts caused the strange behaviour.

Edit: false alarm, the problem comes from elsewhere, it seems to be caused by the subdivision of the terrain tiles.

**Comment by Jeremy-Gaillard** [comment]
The issue seems to happen only when the "spilling" tile is on the verge of being culled by the far plane. I suspect a kind of overflow. The far plan is not well set at the moment, and raising its value pretty much solves the problem, so I don't think this problem is critical.

**Comment by Jeremy-Gaillard** [comment]
Branch is updated and ready to be merged.
Waiting for your validation.

**Comment by gchoqueux** [comment]
I'll check the code before the merger

**Comment by gchoqueux** [comment]
I did not even look at the code, your changes are minor?
Because I think that we should change only the mathematical model: ellipsoid.  #34

**Comment by Jeremy-Gaillard** [comment]
Changes to the existing code are pretty minimal.
BasicMaterial was slightly modified, adding the possibility to set a texture.
c3DEngine was modified to allow initialisation with a 2D camera (needs cleaning)
Scene is used to initialise the flat mode (via the flat boolean)
Most importantly, at l.84 of nodeProcess, the _l_ variable was set to 0 because it made the application bug and I couldn't find the justification for it (cf. comment).

The rest of the code is composed of new classes for the flat mode (which are still improvable, but currently functional).

**Comment by gchoqueux** [comment]
- if 2D camera needs cleaning, it's possible to clean the code?
- l.84 of nodeProcess, what is the bug reference?

**Comment by Jeremy-Gaillard** [comment]
- I use ThreeJS's orbit camera which is not optimal for our kind of application. I lack time to implement a proper camera, unfortunately
- Here's the code I'm refering to:

``` javascript
var l  = 0 * node.absoluteCenter.length(); // TODO : is this really necessary? (note : bugs on plane mode if not 0)
obb.translateZ(l);
obb.update();
var quadInv = obb.quadInverse().clone();
// position in local space
this.camera.setPosition(obb.worldToLocal(camera.position().clone()));
// rotation in local space
this.camera.setRotation(quadInv.multiply(camera.camera3D.quaternion));
obb.translateZ(-l);
```

The bug only happens when in flat mode. In globe mode, node.absoluteCenter.length() is always equal to zero. In flat mode, it is not the case, since the skirt of the tile pushes the absolute centre downwards.
I do not understand why this translation is there: either the obb is in world space, so it shouldn't need this translation, either it is in local space and it needs a translation along the three axes.

**Comment by gchoqueux** [comment]
- In globe mode,`node.absoluteCenter.length()` is always equal to zero -> I don't believe
- the skirt use the center to compute the normal, you need to implement new mathematic model (flat ellipsoid) to compute the normal. (and vertex position, culling)
- `node.absoluteCenter` isn't a good idea, I think to remove this.

**Comment by Jeremy-Gaillard** [comment]
- Just checked, and I was wrong, my bad. The globe still works properly without it though...
- I'm not sure a new mathematical model is needed for something as simple as a plane

Regardless, I think we'll need to have distinct node processes for flat mode and globe mode (since flat mode doesn't need horizon culling for instance). This problem will surely be fixed during the core stabilisation.

**Comment by gchoqueux** [comment]
I think that we should main change is  the mathematical model: flat ellipsoid. It's the good way. it's easier. it's possible for you, to take this way and not to use a new tile?

**Comment by Jeremy-Gaillard** [comment]
Bear in mind that we're rarely in lat/long coordinates while working in the flat mode. Most of the time we'll have a local projection system.
After looking at the ellipsoidTileGeometry code, it seems to be a lot of trouble for little gain. Plus, it prevents some optimisations (less vertices to represent a square than an ellipsoid section for example).

**Comment by gchoqueux** [comment]
A new mathematical model (flat ellipsoid) allows to have the same provider, tile and browser.
This design is more generic and is still compatible. The new enhancements ( #57, #56, #54 ) make your code obsolete.

**Comment by Jeremy-Gaillard** [comment]
It's not so easy.
Part of the code is based on the assumption that we are working with latitude/longitude:
- https://github.com/iTowns/itowns2/blob/master/src/Globe/EllipsoidTileGeometry.js#L99
- https://github.com/iTowns/itowns2/blob/master/src/Globe/EllipsoidTileGeometry.js#L116
- UV2 don't make sense for a plane
  And I maintain my point about the sub-optimality of the generic solution. We could build a planar tile using only 10 triangles with very little processing by keeping a specialised class.

**Comment by gchoqueux** [comment]
The mathematic model should return buffers for the geometry. Furthermore it will facilitate the buffer caching and computation in worker.

**Comment by gchoqueux** [comment]
I'll prepare the code in ellipsoidTileGeometry

**Comment by vpicavet** [comment]
@gchoqueux @Jeremy-Gaillard 
Hi, any news ? where do we stand here ?
I think this is an important PR and we should try to get it merged as soon as possible, and avoid generating conflicts meanwhile, as it also generates a lot of additional work.

**Comment by gchoqueux** [comment]
I make a generic tile with builder.
You need to implement a new builder for flat tile [builder example](https://github.com/iTowns/itowns2/blob/master/src/Globe/BuilderEllipsoidTile.js)

all functions take
 `params {bbox,zoom,segment,center3D,projected}`
- `bbox` global bbox in your projection system 
- `zoom` zoom tile (WMTS)
- `segment` number of segment in tile
- `projected`  current projection in bbox

the builder needs the following functions:
-  `Prepare(params)` to instantiate  `params.projected`
-  `Center(params)` return absolute center tile in cartesian 3D
-  `VertexPosition(params)` return current vertex position 3D cartesian
-  `VertexNormal(params)` return normal for current vertex
-   `uProjecte(u,params)` u coord tile projecte to bounding box, 
-   `vProjecte(v,params)` v coord tile projecte to bounding box
-   `OBB(params)` return oriented bouding box for the tile

optional function
-  `getUV_1(params)` return uv 1 for current vertex, if it isn't defined the uv1 isn't computed

**Comment by Jeremy-Gaillard** [comment]
I'll update the branch as soon as #54 is resolved. I need to be able to apply a different node process to my flat tiles (no horizon culling, etc.).

**Comment by gchoqueux** [comment]
Exactly, with the next browse tree, you will be able to apply a different node process

**Comment by Jeremy-Gaillard** [comment]
I've resumed my work on this branch, but I still encounter problems. I talk about some of them in #70.

I can't reuse the TileMesh classes in their current state:
- I don't use WMTS coordinates
- The groupelevation array is irrelevant to me (I don't even understand its use actually)

On a side note, shouldn't we build the geometry outside of TileMesh? This way we won't have to bother with caching, geometry construction, builders inside TileMesh.

**Comment by gchoqueux** [comment]
This branch is close for the moment

**Comment by Jeremy-Gaillard** [comment]
Development has moved to the planar branch.

**Pull request closed without merging**